channels_redis/core.py

```diff
@@ -310,22 +310,28 @@ def check_receive_lock(self):
         Sends a message to the entire group.
         """
         assert self.valid_group_name(group), "Group name not valid"
-        # Retrieve list of all channel names
+
         key = self._group_key(group)
+
+        group_send_lua = """
+            local prefix = ARGV[1]
+            local group = ARGV[2]
+            local message = ARGV[3]
+            local expiry = ARGV[4]
+            local channels = redis.call('ZRANGE', group, 0, -1)
+            for i,channel in ipairs(channels) do
+                redis.call('RPUSH', prefix .. channel, message)
+                redis.call('EXPIRE', prefix .. channel, expiry)
+            end
+        """
+
         async with self.connection(self.consistent_hash(group)) as connection:
             # Discard old channels based on group_expiry
             await connection.zremrangebyscore(key, min=0, max=int(time.time()) - self.group_expiry)
-            # Return current lot
-            channel_names = [
-                x.decode("utf8") for x in
-                await connection.zrange(key, 0, -1)
-            ]
-        # TODO: More efficient implementation (lua script per shard?)
-        for channel in channel_names:
-            try:
-                await self.send(channel, message)
-            except ChannelFull:
-                pass
+            await connection.eval(
```

Comment:
You're only sending this on a single connection, so it won't work with a multiple-shard setup (the old code's use of individual `send` calls means it picks the right shard for each destination channel).

As the comment you deleted says, a Lua script would have to be run on each shard after getting all the destination channel names and grouping them by which shard their consistent hash puts them on.

Reply:
righty! ill get that addressed. Do you require tests that test it working over multiple connections?

Reply:
That would be ideal, though getting multiple-Redis tests working in Travis has proved hard in the past (which is why there aren't any right now). You might be able to do it with different database numbers on the single redis host they provide.

Reply:
I will try with different database numbers